File: generated_code/cpp/solution_141_2.cpp
#include <iostream>

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        // Using Floyd's Tortoise and Hare algorithm (Two Pointers)
        // Time Complexity: O(N) where N is the number of nodes in the linked list
        // Space Complexity: O(1) auxiliary space
        
        if (!head || !head->next) {
            return false;
        }
        
        ListNode *slow = head;
        ListNode *fast = head;
        
        while (fast && fast->next) {
            slow = slow->next;          // Move slow pointer by 1 step
            fast = fast->next->next;    // Move fast pointer by 2 steps
            
            // If there is a cycle, the fast pointer will eventually catch up to the slow pointer
            if (slow == fast) {
                return true;
            }
        }
        
        // If fast reaches the end, there is no cycle
        return false;
    }
};

File: generated_code/python/solution_141_2.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # Time Complexity: O(N) - In the worst case, we traverse each node at most twice.
        # Space Complexity: O(1) - Only two pointers are used, achieving constant memory.
        # Algorithm: Floyd's Tortoise and Hare (Two Pointers).
        
        # Edge case: An empty list or a list with only one node cannot have a cycle.
        if not head or not head.next:
            return False
        
        slow = head
        fast = head
        
        # Move 'slow' by 1 step and 'fast' by 2 steps.
        # If there's a cycle, the fast pointer will eventually catch up to the slow pointer.
        # If there's no cycle, the fast pointer will reach the end of the list.
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            
            # If the two pointers meet, a cycle exists.
            if slow == fast:
                return True
                
        return False